Project2015To2017/Definition/AssemblyAttributes.cs

```diff
@@ -1,16 +1,174 @@
+
+		public FileInfo File { get; set; }
+
+		private (AttributeListSyntax attList, AttributeSyntax att) FindAttribute(Type attributeType)
+		{
+			//Find the specified attribute and which attribute list it is in
+
+			var attName = attributeType.Name;
+			var altAttName = AttributeShortName(attName);
+
+			return FileContents
+					.AttributeLists
+					.Select(x =>
+						(
+							attList: x,
+							matchingAtts: x.Attributes
+										   .Where(att => att.Name.ToString() == attName || att.Name.ToString() == altAttName)
+										   .ToImmutableArray()
+						)
+					)
+					.Where(x => x.matchingAtts.Any())
+					.Select(x => (x.attList, att: x.matchingAtts.Single()))
+					.LastOrDefault();
+		}
+
+		public string GetAttribute(Type attributeType)
+		{
+			var att = FindAttribute(attributeType).att;
+
+			//Make the assumption that it just has a single string argument
+			//because all of the attributes we currently look for do have
+			return att?.ArgumentList.Arguments.Single().ToString().Trim('"');
+		}
+
+		public void SetAttribute(Type attributeType, string value)
+		{
+			var att = FindAttribute(attributeType);
+
+			SyntaxList<AttributeListSyntax> newAttLists;
+
+			if (value == null)
+			{
+				if (att.att == null)
+				{
+					return;
+				}
+
+				var newAttList = att.attList.RemoveNode(att.att, SyntaxRemoveOptions.KeepNoTrivia);
+
+				if (newAttList.Attributes.Any())
+				{
+					newAttLists = FileContents.AttributeLists.Replace(att.attList, newAttList);
+				}
+				else
+				{
+					newAttLists = FileContents.AttributeLists.Remove(att.attList);
+				}
+			}
+			else if (att.att == null)
+			{
+				var attList = CreateAttribute(attributeType, value);
+
+				newAttLists = FileContents.AttributeLists.Add(attList);
+			}
+			else
+			{
+				var currentArgs = att.att.ArgumentList;
+				var mainArg = currentArgs.Arguments.First();
+				var newMainArg = mainArg.WithExpression(SyntaxFactory.ParseExpression($"\"{value}\""));
+
+				var newArgs = currentArgs.Arguments.Replace(mainArg, newMainArg);
+
+				var newNode = att.att.WithArgumentList(currentArgs.WithArguments(newArgs));
+
+				var newAttList = att.attList.ReplaceNode(att.att, newNode);
+				newAttLists = FileContents.AttributeLists.Replace(att.attList, newAttList);
+			}
+
+			FileContents = FileContents
+							.WithAttributeLists(
+								newAttLists
+							);
+		}
+
+		private static AttributeListSyntax CreateAttribute(Type attributeType, string value)
+		{
+			var root = (CompilationUnitSyntax) SyntaxFactory
+				.ParseSyntaxTree($"[assembly: {attributeType.Name}(\"{value}\")]")
+				.GetRoot();
+
+			var attList = root.AttributeLists.Single();
+
+			return attList;
+		}
+
+		private string AttributeShortName(string fullName)
+		{
+			return fullName.Substring(0, fullName.Length - 9);
+		}
+
+		private bool Equals(AssemblyAttributes other)
+		{
+			return FileContents.ToFullString() == other.FileContents.ToFullString();
+		}
+
+		public override bool Equals(object obj)
```

Comment:
This is dangerous since you don't override gethashcode which you should. See [here](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0659)